### PR TITLE
Issue 1462/interactive brokers data downloader

### PR DESCRIPTION
**Pull request by benjaminthedonkey**
Add Interactive Broker Download  tool

#### Description
Add Interactive Broker Download  tool

#### Related Issue
https://github.com/QuantConnect/Lean/issues/1462

#### Motivation and Context
IB downloader

#### Requires Documentation Change
No change

#### How Has This Been Tested?
I tested with several Forex pairs 

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:

- [x ] My code follows the code style of this project.
- [x ] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes.  The main component (GetHistory for IB ) is already tested 
- [x ] All new and existing tests passed.
- [x ] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

File: ToolBox/IBDownloader/Program.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    {
+        /// <summary>
+        /// Primary entry point to the program
+        /// </summary>
+        static void Main(string[] args)
+        {
+            if (args.Length != 5)
+            {
+                Console.WriteLine("Usage: IBDownloader SECURITY_TYPE SYMBOLS RESOLUTION FROMDATE TODATE");
+                Console.WriteLine("SECURITY_TYPE = eg Equity/Forex");
```
Comment (StefanoRaggi) [inline ToolBox/IBDownloader/Program.cs]: No need to implement in this PR, is there a particular reason for not supporting Options and Futures as well?

File: ToolBox/IBDownloader/Program.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                }
+
+                // Load settings from config.json
+                var dataDirectory = Config.Get("data-folder", "../../../Data");
+
+                // Create IB Broker Gateway Runner
+                InteractiveBrokersGatewayRunner.StartFromConfiguration();
+
+                // Create an instance of the downloader
+                const string market = Market.FXCM; // new market for IB ?
```
Comment (StefanoRaggi) [inline ToolBox/IBDownloader/Program.cs]: This is the first downloader project supporting multiple asset types: maybe the market should be a command line argument (in addition to security type)?

File: ToolBox/IBDownloader/Program.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                        // Download the data
+                        var symbol = Symbol.Create(ticker, securityType, market);
+
+                        var auxEndDate = startDate.AddDays(maxDays); 
+                        auxEndDate = new DateTime(Math.Min(auxEndDate.Ticks, endDate.Ticks));
+
+                        while (startDate < auxEndDate)
+                        {
+                            var data = downloader.Get(symbol, resolution, startDate, auxEndDate);
+                            var bars = data.Cast<QuoteBar>().ToList();
```
Comment (StefanoRaggi) [inline ToolBox/IBDownloader/Program.cs]: For equities, `data` is of type `IEnumerable<TradeBar>`, does this line throw?

File: ToolBox/IBDownloader/IBDataDownloader.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        public IEnumerable<BaseData> Get(Symbol symbol, Resolution resolution, DateTime startUtc, DateTime endUtc)
+        {
+            if (resolution == Resolution.Tick)
+                throw new NotSupportedException("Resolution not available: " + resolution);
+
+            if (endUtc < startUtc)
+                throw new ArgumentException("The end date must be greater or equal than the start date.");
+
+            var barsTotalInPeriod = new List<QuoteBar>();
+            var barsToSave = new List<QuoteBar>();
```
Comment (StefanoRaggi) [inline ToolBox/IBDownloader/IBDataDownloader.cs]: These two variables are unused and should be removed

File: ToolBox/IBDownloader/IBDataDownloader.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    /// IB Downloader class
+    /// </summary>
+    public class IBDataDownloader : IDataDownloader, IDisposable
+    {
+        private readonly InteractiveBrokersBrokerage _brokerage;
+        private readonly InteractiveBrokersSymbolMapper _symbolMapper = new InteractiveBrokersSymbolMapper();
+
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OandaDataDownloader"/> class
```
Comment (StefanoRaggi) [inline ToolBox/IBDownloader/IBDataDownloader.cs]: Class name should be `IBDataDownloader` in comment

File: ToolBox/IBDownloader/Program.cs (hunk 3 above)
Comment (benjaminthedonkey) [reply ToolBox/IBDownloader/Program.cs]: Stefano,  I'm only able to download FOREX. I think this is due to my IB account limitations ( It's just a DEMO account ).  I rather limit the scope of the tool to something I know it's working. Maybe if I sing up for IB I'll come back and add the rest of the assets. 
thanks.

**Pull request merged**